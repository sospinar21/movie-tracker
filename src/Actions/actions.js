export const addMovies = (movieData) => ({
  type: 'ADD_MOVIES',
  movieData
});

export const addFavorites = (movie) => ({
  type: 'ADD_FAVORITE',
  movie
});

export const login = (user) => ({
  type: 'LOGIN',
  user
})

export const logout = () => ({
  type: 'LOGOUT'
})
