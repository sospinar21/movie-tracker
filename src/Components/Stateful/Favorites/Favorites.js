import React, {Component} from 'react';
import './Favorites.css';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { deleteFavorite } from '../../../Helpers/apiCalls';

export class Favorites extends Component{
  constructor(props) {
    super(props)
  }

convertFavorites = () => {
  const foundFavorites = this.props.favorites.map((favorite) => {
    const matchedMovie = this.props.movies.find(movie => movie.movie_id === favorite.favoriteId)
    return matchedMovie;
  });
  return foundFavorites;
}

displayFavorites = () => {
  const favorites = this.convertFavorites();
  const renderedFavorites = favorites.map((movie, index) => {
    return (
      <div key={index} className='movie-card' >
        <p>{movie.title}</p>
        <img src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`} alt={movie.title}/>
        <button onClick={() => deleteFavorite(this.props.user.id, movie.movie_id)}>Remove</button>
      </div>
    )
  })
  return renderedFavorites;
}

render () {
  return (
    <div className='movies-container scroll'>
      {this.displayFavorites()}
    </div>
  );
} 
};

export const mapStateToProps = (state) => {
  return ({
    user: state.user,
    favorites: state.favorites,
    movies: state.movies
  });
};

export default connect(mapStateToProps)(Favorites);