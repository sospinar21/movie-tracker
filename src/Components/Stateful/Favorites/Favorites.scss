.delete-button {
  height: 28px;
  width: 139px;
  position: relative;
  bottom: 11px;
  background-color: #353434;
  border: none;
  color: #26a566;
  font-size: 15px;
  cursor: pointer;
  outline: none;
  -webkit-transition: 0.3s;
  -o-transition: 0.3s;
  transition: 0.3s;
}