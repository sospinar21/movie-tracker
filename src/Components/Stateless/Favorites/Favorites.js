import React from 'react';
import './Favorites.css';
import { connect } from 'react-redux';
import { retrieveFavorites } from './../../../Helpers/apiCalls'

const Favorites = ({user, favorites}) => {
  const favMovies = retrieveFavorites(user.id);
  const favoritesDisplay = favorites.map((movie, index) => {
    return (
      <div key={index} >
        <p>{movie.title}</p>
        <img src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`} alt={movie.title}/>
      </div>
    );
  });
  return (
    <div>
      {favoritesDisplay}
    </div>
  );
};

const mapStateToProps = (state) => {
  return ({
    favorites: state.favorites,
    user: state.user
  });
};

export default connect(mapStateToProps)(Favorites);