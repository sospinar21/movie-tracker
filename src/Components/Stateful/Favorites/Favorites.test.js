import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import React from 'react';
import { Favorites, mapStateToProps, mapDispatchToProps } from './Favorites';
import { removeFavorite } from '../../../Actions/actions';

describe('Favorites', () => {
  
  let favorites;
  let mockProps;

  beforeEach(() => {

    mockProps = {
      user: {},
      favorites: [{favoriteId: 1}],
      movies: [{movie_id: 4, title: 'Spiderman'}, {movie_id: 1, title:'Batman'}],
      removeFavorite: jest.fn()
    };

    favorites = shallow(<Favorites {...mockProps} />, { disableLifecycleMethods: true });
  });
  
  describe('component', () => {
    it('matches snapshot', () => {
      favorites = renderer.create(<Favorites {...mockProps} />, { disableLifecycleMethods: true }).toJSON();
      expect(favorites).toMatchSnapshot();
    });
  });

  describe('component will receive props', () => {

    it('should call displayFavorites', () => {

      favorites.instance().displayFavorites = jest.fn()
      favorites.instance().componentWillReceiveProps()

      expect(favorites.instance().displayFavorites).toHaveBeenCalled()
    })
  })

  describe('displayFavorites', () => {

    it('should only display matched favorites', () => {

      const expected = [{movie_id: 1, title:'Batman'}]

      expect(favorites.instance().convertFavorites()).toEqual(expected)


    });
  });

  describe('updateFavoritesOnDelete', () => {

    it('calls delete favorite with the correct params', async () => {
      
      const movie = {favoriteId: 1};

      window.fetch = jest.fn();

      favorites.instance().updateFavoritesOnDelete(movie);
      expect(mockProps.removeFavorite).toHaveBeenCalledWith(movie);
    });
  });

  describe('mapStateToProps', () => {

    it('should return an object with user, favorites and movies', () => {

      const mockState = {
        user: {name: 'Tony'},
        favorites: [],
        movies: []
      };

      const expected = {
        user: {name: 'Tony'},
        favorites: [],
        movies: []
      };

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);

    });
  });

  describe('mapDispatchToProps', () => {

    it('should call dispatch with the correct params', () => {

      const mockDispatch = jest.fn();
      const movie = {movie_id: 1, title:'Batman'}
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'REMOVE_FAVORITE',
        movie
      };
      
      mappedProps.removeFavorite(movie);
      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    })
  })
});