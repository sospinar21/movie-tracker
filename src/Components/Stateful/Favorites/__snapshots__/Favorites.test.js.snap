// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites component matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorites={
        Array [
            Object {
              "favoriteId": 1,
            },
          ]
    }
    movies={
        Array [
            Object {
              "movie_id": 4,
              "title": "Spiderman",
            },
            Object {
              "movie_id": 1,
              "title": "Batman",
            },
          ]
    }
    removeFavorite={[Function]}
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="movie-card"
>
          <img
                    alt="Batman"
                    src="https://image.tmdb.org/t/p/w500undefined"
          />
          <button
                    className="delete-button"
                    onClick={[Function]}
          >
                    Remove
          </button>
</div>,
      ],
      "className": "movies-container scroll",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="Batman"
              src="https://image.tmdb.org/t/p/w500undefined"
/>,
            <button
              className="delete-button"
              onClick={[Function]}
>
              Remove
</button>,
          ],
          "className": "movie-card",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Batman",
              "src": "https://image.tmdb.org/t/p/w500undefined",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Remove",
              "className": "delete-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Remove",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="movie-card"
>
            <img
                        alt="Batman"
                        src="https://image.tmdb.org/t/p/w500undefined"
            />
            <button
                        className="delete-button"
                        onClick={[Function]}
            >
                        Remove
            </button>
</div>,
        ],
        "className": "movies-container scroll",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="Batman"
                src="https://image.tmdb.org/t/p/w500undefined"
/>,
              <button
                className="delete-button"
                onClick={[Function]}
>
                Remove
</button>,
            ],
            "className": "movie-card",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Batman",
                "src": "https://image.tmdb.org/t/p/w500undefined",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Remove",
                "className": "delete-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Remove",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
